packages/dds/shared-object-base/src/test/attachingBindingAndConnecting.spec.ts

```diff
@@ -0,0 +1,493 @@
+	? T
+	: {
+			-readonly [P in keyof T]?: Overridable<T[P]>;
+	  };
+
+function createOverridableProxy<T extends object>(name: string, ...overrides: Overridable<T>[]) {
+	return new Proxy<T>({} as any as T, {
+		get: (_, p, r) => {
+			for (const override of overrides) {
+				if (p in override) {
+					return Reflect.get(override, p, r);
+				}
+			}
+			assert.fail(`No override for ${name}.${p.toString()}`);
+		},
+	});
+}
+
+function createTestSharedObject(
+	overrides: Overridable<{
+		id: string;
+		runtime: IFluidDataStoreRuntime;
+		attributes: IChannelAttributes;
+		telemetryConfigPrefix: string;
+		summarizeCore: (
+			this: SharedObject,
+			serializer: IFluidSerializer,
+			telemetryContext?: ITelemetryContext | undefined,
+			incrementalSummaryContext?: IExperimentalIncrementalSummaryContext | undefined,
+		) => ISummaryTreeWithStats;
+		loadCore: (this: SharedObject, services: IChannelStorageService) => Promise<void>;
+		processCore: (
+			this: SharedObject,
+
+			message: ISequencedDocumentMessage,
+			local: boolean,
+			localOpMetadata: unknown,
+		) => void;
+		onDisconnect: (this: SharedObject) => void;
+		applyStashedOp: (this: SharedObject, content: any) => unknown;
+		didAttach: () => void;
+	}>,
+) {
+	class TestSharedObject extends SharedObject {
+		protected summarizeCore = overrides?.summarizeCore?.bind(this);
+		protected loadCore = overrides?.loadCore?.bind(this);
+		protected processCore = overrides?.processCore?.bind(this);
+		protected onDisconnect = overrides?.onDisconnect?.bind(this);
+		protected applyStashedOp = overrides?.applyStashedOp?.bind(this);
+		protected didAttach =
+			overrides.didAttach?.bind(this) ?? (() => assert.fail("didAttach not set"));
+	}
+
+	const runtime = overrides?.runtime ?? {};
+	runtime.channelsRoutingContext ??= {
+		absolutePath: "/",
+		attachGraph: () => {},
+		get IFluidHandleContext() {
+			return this;
+		},
+		isAttached: false,
+		resolveHandle: async () => ({ status: 500, mimeType: "error", value: "error" }),
+	};
+	runtime.IFluidHandleContext ??= runtime.channelsRoutingContext;
+
+	runtime.logger ??= createChildLogger();
+
+	const attributes = overrides?.attributes ?? {};
+
+	attributes.type ??= "TestSharedObject";
+
+	return {
+		overrides,
+		sharedObject: new TestSharedObject(
+			overrides?.id ?? "testSharedObject",
+			createOverridableProxy<IFluidDataStoreRuntime>(
+				"runtime",
+				runtime,
+				new TypedEventEmitter<IFluidDataStoreRuntimeEvents>() as any as Overridable<IFluidDataStoreRuntime>,
+			),
+			createOverridableProxy<IChannelAttributes>("attributes", attributes),
+			overrides?.telemetryConfigPrefix ?? "testSharedObject",
+		),
+	};
+}
+
+const attachStateAndConnectedMatrix = generatePairwiseOptions({
+	connected: [true, false],
+	attachState: [AttachState.Detached, AttachState.Attaching, AttachState.Attached],
+});
+
+describe("SharedObject attaching binding and connecting", () => {
```

Comment:
i'm generally ok with the setup of these tests, but it's difficult to tell what tests are trying to test and why based on their name or other top-level comments in the test suite. E.g. the PR description is actually more informative for that matter:

> This was conceptually and functionally incorrect, as the attach state of a dds is determined by two facts:
> - Is it bound to a context which means its handle is stored in an already bound dds
> - Is the runtime attached
> if both of these are true, then the dds should be considered attached.
> Beyond attach, there is also the connected state, and the didAttach method. The connect state should always be false while detached, and didAttach should only be called once when the dds transitions.
> This change tightens up the handling of the bind, attached, and connects state of shared object while decoupling it from the services, to allow the services to be set sooner. Lastly, this change adds a ton of previously missing unit test coverage to ensure all the states are managed properly.

Putting some of that language here--or better, on the appropriate SharedObject methods in a more precise fashion and referencing the invariants here--would probably help add clarity here.

Reply:
i'll add more comments!